api/internal/plugins/loader/loader.go

```diff
@@ -47,8 +47,18 @@ func (l *Loader) Config() *types.PluginConfig {
 	return l.pc
 }
 
-// SetWorkDir sets the working directory for this loader's plugins
-func (l *Loader) SetWorkDir(wd string) {
+// DeepCopyPluginConfig makes a full copy the actual values of PluginConfig.
+func (l *Loader) DeepCopyPluginConfig() {
```

Comment:
I'm behind this implementation because it fixes issue #4347. Ideally, though, we'd have a solution that's easier to understand. 

Need @KnVerey's feedback on this, but what about writing a 
```
func (l *Loader) CopyLoader() *Loader
``` 
method that only copies the relevant fields?

For instance, we 
* shouldn't make a deep copy of `fs filesys.FileSystem` because we want our `Loader`s to run on the same file system
* should copy `rf *resmap.Factory` because we [don't make a copy of it in the factory](https://github.com/kubernetes-sigs/kustomize/blob/master/api/internal/plugins/loader/loader.go#L40) and [expose it again here](https://github.com/kubernetes-sigs/kustomize/blob/master/api/internal/plugins/loader/loader.go#L213). We can do so by defining a `Copy()` method on `resmap.Factory` that we call.
* should definitely copy `pc *types.PluginConfig`, by defining `Copy()` methods at the `types.PluginConfig` and `FnpLoadingOptions` levels.

Another easier, though less safe, alternative is to store `struct`s instead of pointers to them. For example, in `Loader` we can store `pc types.PluginConfig` instead of `pc *types.PluginConfig` and thus we can use the value copy that go gives us. This wouldn't give us a complete copy, but would fix the krm issue.

Reply:
That makes sense :+1: 

But if we agreed on the fact that it's not a deep copy ... and follow the same logic, isn't calling it `copy` or `new` would also be confusing? :thinking: 

Because the `Loader` struct fields have some pointers. Which is the root cause for the issue here (the nested cross-reference between copies).

And for the same reason (the `Loader` struct uses pointers), we cannot make it like the one in `fileloader` because we need to update a nested value in the struct.

So my suggestion as in the commit [proposal v2](https://github.com/kubernetes-sigs/kustomize/pull/4654/commits/05a752fba4c6b7c5590bb349e7c4c2947860cd8c) where it returns the loader after setting the working dir.

It's not `copy` and not `new` because, as mentioned, the cross-referencing.

Please correct me if I got it wrong.

Reply:
I think your current `LoaderWithWorkingDir` solution implements @KnVerey's 2nd suggestion spot on.